pkg/certificate/auto/provider.go

```diff
@@ -0,0 +1,297 @@
+
+func (ac *Provider) GetCertificateConfig(ctx context.Context,
+	secretName, namespace string) (*interfaces.Config, error) {
+	var autoCertSecret corev1.Secret
+	err := ac.Get(ctx, client.ObjectKey{Name: secretName, Namespace: namespace}, &autoCertSecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get certificate: %w", err)
+	}
+
+	return &interfaces.Config{}, nil
+}
+
+// parsePrivateKey parses the private key from the PEM-encoded data.
+func parsePrivateKey(privateKeyData []byte) (crypto.PrivateKey, error) {
+	block, _ := pem.Decode(privateKeyData)
+	if block == nil {
+		return nil, errors.New("failed to decode private key: invalid PEM")
+	}
+
+	// Parse the private key from the PEM block
+	privateKey, err := x509.ParseECPrivateKey(block.Bytes)
+	if err != nil {
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse private key: %w", err)
+		}
+	}
+
+	return privateKey, nil
+}
+
+// checkKeyPair checks if the private key matches the certificate by validating the public key
+func checkKeyPair(cert *x509.Certificate, privateKey crypto.PrivateKey) error {
+	switch key := privateKey.(type) {
+	case *rsa.PrivateKey:
+		pub, ok := cert.PublicKey.(*rsa.PublicKey)
+		if !ok || !key.PublicKey.Equal(pub) {
+			return interfaces.ErrRSAKeyPair
+		}
+	case *ecdsa.PrivateKey:
+		pub, ok := cert.PublicKey.(*ecdsa.PublicKey)
+		if !ok || !key.PublicKey.Equal(pub) {
+			return interfaces.ErrECDSAKeyPair
+		}
+	case *ed25519.PrivateKey:
+		pub, ok := cert.PublicKey.(ed25519.PublicKey)
+		if !ok || !bytes.Equal(key.Public().(ed25519.PublicKey), pub) {
+			return interfaces.ErrED25519KeyPair
+		}
+	default:
+		return fmt.Errorf("unsupported private key type: %T", key)
+	}
+
+	return nil
+}
+
+// createNewSecret generates or updates a Kubernetes TLS Secret
+// with a self-signed certificate in the specified namespace.
+// DNSNames and IPAddresses if not user-defined, will be set to default value in runtime:
+// fmt.Sprintf("%s-%d.%s.%s.svc.cluster.local", ec.Name, index, ec.Name, ec.Namespace)
+// minimum validityDuration is 365 days
+// https://github.com/etcd-io/etcd/blob/b87bc1c3a275d7d4904f4d201b963a2de2264f0d/client/pkg/transport/listener.go#L275
+func (ac *Provider) createNewSecret(ctx context.Context, secretName, namespace string,
+	cfg *interfaces.Config) error {
+	validity := DefaultValidity
+	if cfg.ValidityDuration != 0 {
+		validity = cfg.ValidityDuration * time.Hour
+	}
+
+	tmpDir, err := os.MkdirTemp("", fmt.Sprintf("etcd-auto-cert-%s-*", secretName))
+	if err != nil {
+		return fmt.Errorf("failed to create temp dir for certificate generation: %w", err)
+	}
+
+	defer func() {
+		_ = os.RemoveAll(tmpDir)
+	}()
+
+	var hosts []string
+	if cfg != nil {
+		if cfg.CommonName != "" {
+			hosts = append(hosts, cfg.CommonName)
+		}
+		if len(cfg.AltNames.DNSNames) > 0 {
+			hosts = append(hosts, cfg.AltNames.DNSNames...)
+		}
+	}
+
+	tlsInfo, selfCertErr := transport.SelfCert(zap.NewNop(), tmpDir, hosts, uint(validity/DefaultValidity))
```

Comment:
if you decide to use the fake host ports as we discussed during the meeting of 25 Nov, due to the bug in SelfCert not working for hosts without port, then remember to use net.JoinHostPort for adding a port. the function is IPv4/6 compatible.